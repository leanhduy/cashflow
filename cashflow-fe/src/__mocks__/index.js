export { mockFinancialStatusProps, mockIncomes } from './mockProps'
