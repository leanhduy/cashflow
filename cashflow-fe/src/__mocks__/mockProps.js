export const mockFinancialStatus = {
  totalIncome: 2000,
  totalExpenses: 1280,
  cash: 670,
}

export const mockIncomes = [
  {
    id: 1,
    name: 'Programmer Salary',
    amount: 2500,
    type: 'salary',
  },
  {
    id: 2,
    name: '3/2 House',
    amount: 100,
    type: 'estate',
  },
  {
    id: 3,
    name: '2/1 Condo',
    amount: 2500,
    type: 'estate',
  },
  {
    id: 4,
    name: '20 Shares of 2BIG',
    amount: 200,
    type: 'dividends',
  },
  {
    id: 5,
    name: '100 Shares of CD',
    amount: 2000,
    type: 'dividends',
  },
  {
    id: 6,
    name: '100 Shares of AB',
    amount: 4000,
    type: 'dividends',
  },
]

export const mockAssets = [
  {
    id: 1,
    name: 'Shares of 2BIG',
    quantity: 5,
    value: 1200,
    type: 'Stocks/Funds/CDs',
  },
  {
    id: 2,
    name: 'Shares of ON2U',
    quantity: 10000,
    value: 5,
    type: '',
  },
  {
    id: 3,
    name: 'Startup',
    quantity: 1,
    value: 3000,
    type: 'Real Estate/Business',
  },
  {
    id: 4,
    name: 'Gold Coin',
    quantity: 1,
    value: 3000,
    type: 'Real Estate/Business',
  },
  {
    id: 5,
    name: '2/1 CONDO',
    quantity: 1,
    value: 60000,
    type: 'Real Estate/Business',
  },
  {
    id: 6,
    name: '3/2 HOUSE',
    quantity: 1,
    value: 50000,
    type: 'Real Estate/Business',
  },
  {
    id: 7,
    name: 'PLEX',
    quantity: 1,
    value: 200000,
    type: 'Real Estate/Business',
  },
]

export const mockChildExpense = {
  childNum: 1,
  expensePerChild: 200,
}

export const mockLiabilities = [
  {
    id: 1,
    name: 'Home Mortgage',
    value: 50000,
    type: 'home',
  },
  {
    id: 2,
    name: 'Car Loans',
    value: 3000,
    type: 'car',
  },
  {
    id: 3,
    name: 'Credit Cards',
    value: 3000,
    type: 'credit',
  },
  {
    id: 4,
    name: 'Retail Debt',
    value: 1000,
    type: 'retail',
  },
  {
    id: 5,
    name: 'Loan',
    value: 12000,
    type: 'bank',
  },
]

export const mockStartDialog = {
  playerName: 'Player',
  description: 'When you are ready, roll the dice and take your turn',
  note: 'Before you start your turn, review your financial statement. You may also use this time to repay liabilities or borrow money.',
}

export const mockRepayDialog = {
  title: 'PAY OFF LOAN',
  description: 'Select a liability on your statement sheet to repay your debt.',
  liabilities: mockLiabilities,
}

export const mockDownsizedDialog = {
  title: 'DOWNSIZED!',
  description: 'Pay a full set of your expenses and lose two turns and charity',
  expenses: 1500,
}

export const mockBabyDialogNew = {
  title: 'NEW BABY!',
  description: 'Congratulations! One child has been added to your dependents',
  expenses: 110,
}

export const mockBabyDialogLimitReached = {
  title: 'BABY LIMIT REACHED!',
  description: 'You cannot have more than 3 child',
  expenses: 0,
}

export const mockOpportunityDialog = {
  title: 'DEAL OPPORTUNITY',
  description: 'Which deal type do you want?',
  info: 'Small deals cost $5,000 or less. Big deals cost $6,000 or more.',
}

export const mockSmallDeal = {
  id: 1,
  title: 'MUTUAL FUND - GRO4US FUND',
  description:
    'Weak earnings by most companies lead to weak price of mutual fund.',
  note: 'Trading Range: $10 to $30',
  type: 'stock',
  arg1: 10, // cost
  arg2: 0, // cashflow
  arg3: 10, // tradingMin
  arg4: 30, // tradingMax
}

export const mockBigDeal = {
  id: 1,
  title: '8-plex for Sale',
  description:
    'Professional person urgently needs cash to save partnership. 8-plex sale to raise capital, good opportunity for right person.',
  type: 'real-estate',
  arg1: 160000, // cost
  arg2: 1700, // cashflow
  arg3: 32000, // downpay
  arg4: 128000, // mortgage
}

export const mockMarket = {
  id: 1,
  title: 'APARTMENT HOUSE BUYER',
  description:
    'You are offered $25,000 per unit for all units in apartment houses of any size. Buyer has their own financing. If you sell, pay off the related mortgage and give up the cash flow you currently receive on this property.',
  type: 'real-estate',
  price: 25000, // selling cost
}

export const mockDoodad = {
  id: 1,
  title: 'GO OUT TO DINNER',
  description: 'Pay $80',
  cost: -80, // selling cost
}
