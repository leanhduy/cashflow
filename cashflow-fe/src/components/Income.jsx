import styled from '@emotion/styled'
import { colors } from '@/styles'
import PropTypes from 'prop-types'
import { currencyFormatter } from '@/utils/helpers'

const Income = ({ incomes }) => {
  return (
    <CardContainer>
      <CardHeader>INCOME</CardHeader>
      <CardBody>
        <ListHeader>Cash Flow</ListHeader>
        <StyledList>
          {incomes &&
            incomes.map((i) => (
              <ListItem key={i.id}>
                <ListItemLeft>{i.name}</ListItemLeft>
                <ListItemRight>
                  ${currencyFormatter.format(i.amount)}
                </ListItemRight>
              </ListItem>
            ))}
        </StyledList>
      </CardBody>
    </CardContainer>
  )
}

//#region prop types
Income.propTypes = {
  incomes: PropTypes.array,
}
//#endregion prop types

export default Income

//#region styled components
const CardContainer = styled.div({
  backgroundColor: colors.white,
  border: `1px solid ${colors.purple.base}`,
  borderRadius: '12px',
  maxWidth: '400px',
  margin: 0,
  padding: 0,
  height: '100%',
  display: 'flex',
  flexDirection: 'column',
})

const CardHeader = styled.h3({
  backgroundColor: colors.green.base,
  borderRadius: '10px 10px 0 0',
  color: colors.white,
  fontWeight: 700,
  height: '1.5rem',
  margin: 0,
  padding: '.25rem',
  textAlign: 'center',
})

const CardBody = styled.div({
  margin: '.25rem 1rem 0',
  paddingRight: '1rem',
  height: '70%',
})

const StyledList = styled.ul({
  overflowY: 'scroll',
  height: '100%',
  scrollbarWidth: 'none',
  paddingInlineStart: 0,
  margin: 0,
})

const ListHeader = styled.div({
  textAlign: 'right',
  fontSize: '.875rem',
  margin: 0,
  fontWeight: 800,
  padding: 0,
  borderBottom: `1px solid ${colors.purple.light}`,
})

const ListItem = styled.li({
  display: 'flex',
  fontSize: '.875rem',
  flexDirection: 'row',
  justifyContent: 'space-between',
  padding: '.375rem 0',
  columnGap: '10px',
})

const ListItemLeft = styled.span({
  borderBottom: `1px solid ${colors.purple.light}`,
  flex: '1 1 150px',
  fontSize: '.875rem',
  textAlign: 'left',
})
const ListItemRight = styled.span({
  borderBottom: `1px solid ${colors.purple.light}`,
  flex: '1 1 auto',
  fontSize: '.875rem',
  textAlign: 'right',
})
//#endregion styled components
