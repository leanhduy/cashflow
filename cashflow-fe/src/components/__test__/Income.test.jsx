import { screen } from '@testing-library/dom'
import { render } from '@testing-library/react'
import { describe, expect, it, vi, beforeEach } from 'vitest'
import { mockIncomes } from '@/__mocks__'
import Income from '../Income'

describe('Income component', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  it('should render correct elements for a Income component', () => {
    render(<Income incomes={mockIncomes} />)
    const title = screen.getByRole('heading')
    const listheader = screen.getByText(/Cash Flow/)
    expect(title).toBeInTheDocument()
    expect(listheader).toBeInTheDocument()
    expect(title.textContent).to.equal('INCOME')
  })

  it('should render correct number of incomes', () => {
    render(<Income incomes={mockIncomes} />)
    const listItems = screen.getAllByRole('listitem')
    expect(listItems.length).to.equal(mockIncomes.length)
  })

  it('should render correctly when no incomes are provided', () => {
    render(<Income />)
    const listItems = screen.queryAllByRole('listitem')
    expect(listItems.length).to.equal(0)
  })

  // > Return to this one later, when we fully know how to mock module
  // it('should format income amount correctly', () => {
  //   render(<Income incomes={mockIncomes} />)
  //   mockIncomes.forEach((income) => {
  //     expect(currencyFormatter.format).toHaveBeenCalledWith(income.amount)
  //   })
  // })

  // ? This technique is called 'snapshot testing'
  // ? Purpose: Capture the rendered output and detect any unintended changes in the component over time.
  it('should match the snapshot', () => {
    const { asFragment } = render(<Income incomes={mockIncomes} />)
    expect(asFragment()).toMatchSnapshot()
  })
})
