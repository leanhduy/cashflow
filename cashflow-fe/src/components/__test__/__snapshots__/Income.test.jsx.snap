// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Income component > should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1l7yyhn"
  >
    <h2
      class="css-ybmwtq"
    >
      INCOME
    </h2>
    <div
      class="css-7xvfqo"
    >
      <div
        class="css-kcy40u"
      >
        Cash Flow
      </div>
      <ul
        class="css-8707tt"
      >
        <li
          class="css-bf19dc"
        >
          <p>
            Programmer Salary
          </p>
          <p>
            $2,500
          </p>
        </li>
        <li
          class="css-bf19dc"
        >
          <p>
            3/2 House
          </p>
          <p>
            $100
          </p>
        </li>
        <li
          class="css-bf19dc"
        >
          <p>
            2/1 Condo
          </p>
          <p>
            $2,500
          </p>
        </li>
        <li
          class="css-bf19dc"
        >
          <p>
            20 Shares of 2BIG
          </p>
          <p>
            $200
          </p>
        </li>
        <li
          class="css-bf19dc"
        >
          <p>
            100 Shares of CD
          </p>
          <p>
            $2,000
          </p>
        </li>
        <li
          class="css-bf19dc"
        >
          <p>
            100 Shares of AB
          </p>
          <p>
            $4,000
          </p>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
